src/ir/item.rs

```diff
@@ -38,13 +38,46 @@ pub trait ItemCanonicalName {
 /// }
 /// ```
 ///
-/// For bar, the canonical path is `foo::BAR`, while the canonical name is just
-/// `BAR`.
+/// For bar, the canonical path is `vec!["foo", "BAR"]`, while the canonical
+/// name is just `"BAR"`.
 pub trait ItemCanonicalPath {
     /// Get the canonical path for this item.
     fn canonical_path(&self, ctx: &BindgenContext) -> Vec<String>;
 }
 
+/// A trait for iterating over an item and its parents and up its ancestor chain
+/// up to (but not including) the implicit root module.
+pub trait ItemAncestors {
+    /// Get an iterable over this item's ancestors.
+    fn ancestors<'a, 'b, 'c>(&'a self,
```

Comment:
Wouldn't the former restrict `'a` to the intersection of `self`'s borrow and the `BindgenContext`'s borrow, which would basically just be `self`'s borrow? The iterator could safely outlive the item, it just cannot outlive the context.

Regarding the latter, I was running into issues in the `WhitelistedItemsIter` when I tried to use one lifetime for both the reference to the context and the context's internal `GenContext` lifetime, that ended up preventing compilation. Because of that, I just skipped right to using different lifetime parameters here.